docs/core/v3.0.rst

```diff
@@ -1031,10 +1031,11 @@ following mandatory names:
     value must be a number with no fraction or exponent part and must
     be within the range of the data type.
 
-    For any data type, if the ``fill_value`` is the literal ``null``
-    then the fill value is undefined and the implementation may use
-    any arbitrary value that is consistent with the data type as the
-    fill value.
+    For any data type, the ``fill_value`` is required. The literal
+    ``null`` is not permitted. The fill value needs to be defined
```

Comment:
wouldn't the literal `null` be a permitted fill value for an array of objects? Maybe just remove the sentence which bars the literal null.

Reply:
I agree that it doesn't make sense to specifically exclude `null`.

Reply:
In Zarr Python, arrays of objects already need an object codec and go through encoding before writing. If we are already handling encoding with fill values (thinking about `base64` discussion), we could expect that happens here (even for `None`).

That said, the original spec didn't cover object types. We may need to do work on [the object type proposal]( https://github.com/zarr-developers/zarr-specs/blob/6b66d484cbf985221dc0badb003b2b3937214d5a/docs/extensions/data-types/object/v1.0.rst#editors-draft-2-march-2022 ) to cover this (cc @jstriebel)

Reply:
Agreed that object data types require further discussion (see #146).  But since the allowed values for fill_value are already specified I don't think it is necessary to specifically prohibit null.  For a hypothetical json data type, it could make sense to encode the fill value as a literal json value and in that case null would be valid.

Reply:
I agree that it should be unnecessary for exclude `null` since it is already implicitly excluded by the previous sentences. However, given that `null` fill_values are abundant in v2 I think it is worth mentioning here. Maybe we can modify the sentence to something like: This means that in contrast to v2 of this spec ``null`` is not permitted as a fill_value for numeric data types.

Reply:
The reason to suggest banning `null` is to handle the issue we encountered when trying to skip writing empty chunks. Namely using `null` for `fill_value` is underspecified. One alternative (if we want to solve the write optimization problem) might be to better specify what `null` means

Reply:
I agree that it's easier to allow null later if needed, and that it's mostly emphasizing that it's not somehow optional, not even by putting in `null`. If `null` becomes a viable value for some codecs, let's add it then.